### PR TITLE
Fixed python error message concurrency issue, add a button to force kill the running program

### PR DESCRIPTION
**Pull request by ruiqizhu26**

### CODE REVIEW

File: basestation/base_station_interface.py [hunk 1]
```diff
@@ -148,14 +150,8 @@ def post(self):
             if self.send_blockly_remote_server:
                 url = 'http://127.0.0.1:5000/code/'
```
Comment (virenvshah) [inline basestation/base_station_interface.py:151]: The Flask Migration project will get rid of these hard-coded URLs :)

File: basestation/base_station_interface.py [hunk 2]
```diff
@@ -280,22 +277,36 @@ def get(self):
 class ErrorMessageHandler(tornado.websocket.WebSocketHandler):
```
Comment (virenvshah) [inline basestation/base_station_interface.py:277]: Add doc-string please

File: basestation/base_station_interface.py [hunk 3]
```diff
@@ -280,22 +277,36 @@ def get(self):
...
+        elif error_message == "Successful execution":
             error_json = {"error": error_message, "code": 1}
         else:
             error_json = {"error": error_message, "code": 0}
-        print("error_json is: ")
-        print(error_json)
         self.write(json.dumps(error_json).encode())
 
 
+class StopHandler(tornado.websocket.WebSocketHandler):
```
Comment (virenvshah) [inline basestation/base_station_interface.py:296]: Why do we have a separate StopHandler?

File: basestation/bot/pi_bot.py [hunk 4]
```diff
@@ -78,6 +78,7 @@ def tcp_parse_incoming(self, data):
             """
             parse incoming data
             """
+            print("REACHES TCP_PARSE_INCOMING")
```
Comment (virenvshah) [inline basestation/bot/pi_bot.py:81]: Remove this?

File: minibot/new_minibot.py [hunk 5]
```diff
@@ -54,6 +59,7 @@ def parse_command(cmd, tcpInstance):
     # our code execution pointer would get stuck in the infinite loop.
     if key == "WHEELS":
         if value == "forward":
+            print("Pam is moving forward???")
```
Comment (virenvshah) [inline minibot/new_minibot.py:62]: Remove this

File: minibot/new_minibot.py [hunk 6]
```diff
@@ -88,12 +94,14 @@ def parse_command(cmd, tcpInstance):
...
                 file.write(program)
                 file.close()
-                return_value = spawn_script_process(script_name)
-                return return_value
+                spawn_script_process(script_name, tcpInstance)
             except Exception as exception:
-                print("Exception occurred at compile time")
-                str_exception = str(type(exception)) + ": " + str(exception)
-                return str_exception
+                print("Exception occurred")
```
Comment (virenvshah) [inline minibot/new_minibot.py:99]: Remove this?

File: minibot/new_minibot.py [hunk 7]
```diff
@@ -115,21 +123,27 @@ def process_string(value):
...
     Args:
         scriptname (:obj:`str`): The name of the script to run.
     """
     time.sleep(0.1)
 
-    with concurrent.futures.ThreadPoolExecutor() as executor:
-        future = executor.submit(run_script, scriptname)
-        return_value = future.result()
-        return return_value
+    # with concurrent.futures.ThreadPoolExecutor() as executor:
```
Comment (virenvshah) [inline minibot/new_minibot.py:134]: Remove the commented code

File: minibot/new_minibot.py [hunk 8]
```diff
@@ -115,21 +123,27 @@ def process_string(value):
     return program
 
 
-def spawn_script_process(scriptname):
+def spawn_script_process(scriptname, tcpInstance):
```
Comment (virenvshah) [inline minibot/new_minibot.py:126]: Hmm so we had to pass the tcpInstance in the end?  There was no way around it?

File: minibot/new_minibot.py [hunk 9]
```diff
@@ -115,21 +123,27 @@ def process_string(value):
...
-        return_value = future.result()
-        return return_value
+    # with concurrent.futures.ThreadPoolExecutor() as executor:
+    #     future = executor.submit(run_script, scriptname)
+    #     return_value = future.result()
+    #     return return_value
+    stop == False
+    manager = Manager()
+    output = manager.Value(c_char_p, "")
+    global current_process
```
Comment (virenvshah) [inline minibot/new_minibot.py:141]: Let's declare the globals at the top of the function

File: minibot/new_minibot.py [hunk 10]
```diff
@@ -115,21 +123,27 @@ def process_string(value):
...
 
-    with concurrent.futures.ThreadPoolExecutor() as executor:
-        future = executor.submit(run_script, scriptname)
-        return_value = future.result()
-        return return_value
+    # with concurrent.futures.ThreadPoolExecutor() as executor:
+    #     future = executor.submit(run_script, scriptname)
+    #     return_value = future.result()
+    #     return return_value
+    stop == False
```
Comment (virenvshah) [inline minibot/new_minibot.py:138]: Lol what is this?

File: minibot/new_minibot.py [hunk 11]
```diff
@@ -115,21 +123,27 @@ def process_string(value):
...
+    manager = Manager()
+    output = manager.Value(c_char_p, "")
+    global current_process 
+    current_process = Process(target=run_script, args=(scriptname, tcpInstance))
+    current_process.start()
 
 
-def run_script(scriptname):
+def run_script(scriptname, tcp_instance):
```
Comment (virenvshah) [inline minibot/new_minibot.py:146]: Do we need both the run_script and the spawn_script_process functions?

File: minibot/scripts/bot_script.py [hunk 12]
```diff
@@ -2,4 +2,5 @@
 import time
 from threading import *
 def run():
-    1/0hjfjgdsgdy
```
Comment (virenvshah) [inline minibot/scripts/bot_script.py:5]: Don't push this file, could you add this file to the .gitignore?

File: static/gui/static/js/components/AddBot.js [hunk 13]
```diff
@@ -214,6 +215,23 @@ export default class AddBot extends React.Component {
             })
     }
 
+    killPythonListener() {
```
Comment (virenvshah) [inline static/gui/static/js/components/AddBot.js:218]: Can we make sure the regular Stop button also kills the programs?

File: static/gui/static/js/components/blockly.js [hunk 14]
```diff
@@ -157,31 +159,76 @@ class PythonTextBox extends React.Component {
...
-          document.getElementById("errormessage").style.color="#32CD32";
-        }
-        else {
-          // red
-          document.getElementById("errormessage").style.color="#FF0000";
-        }
-      })
-      .catch((err) => {
-        console.log(err)
+    let interval = setInterval(function() {
```
Comment (virenvshah) [inline static/gui/static/js/components/blockly.js:162]: could you add a brief comment describing what the setInterval does?

File: static/gui/static/js/components/blockly.js [hunk 15]
```diff
@@ -157,31 +159,76 @@ class PythonTextBox extends React.Component {
...
+          document.getElementById("errormessage").value = response.data["error"];
+          // if the code is -1 it means the result hasn't arrived yet, hence 
+          // we shouldn't clear the interval and should continue polling
+          if (response.data["code"] !== -1) {
+            if (response.data["code"] === 1) {
+              // lime green 
+              document.getElementById("errormessage").style.color="#32CD32";
+            }
+            else {
+            // red
```
Comment (virenvshah) [inline static/gui/static/js/components/blockly.js:180]: could you indent the comment

File: static/gui/static/js/components/blockly.js [hunk 16]
```diff
@@ -157,31 +159,76 @@ class PythonTextBox extends React.Component {
...
+        })
+        .catch((err) => {
+          console.log(err)
+        })
+      }, 500);
+  }
 
+  stop(event) {
+    console.log('sent stop command');
+    // axios({
```
Comment (virenvshah) [inline static/gui/static/js/components/blockly.js:196]: delete the commented code

**Review by virenvshah — CHANGES_REQUESTED** [review]

File: minibot/new_minibot.py [hunk 17]
```diff
@@ -9,6 +9,8 @@
 import ast
```
Comment (virenvshah) [inline minibot/new_minibot.py:9]: do we ever use this module?

Comment (virenvshah) [reply minibot/new_minibot.py:9]: oh actually I think we might be using it for creating the bot_script.py program file.

**Comment by virenvshah** [comment]
Could you also merge master into your branch so we resolve all merge conflicts?

File: minibot/new_minibot.py (hunk 8 above)
Comment (ruiqizhu26) [reply minibot/new_minibot.py:126]: It seems like a function run on another process cannot change global variables, so this is probably the only way?

Comment (ruiqizhu26) [reply minibot/new_minibot.py:126]: (I ran a little experiment myself)

File: minibot/scripts/bot_script.py [hunk 18]
```diff
@@ -2,5 +2,4 @@
 import time
```
Comment (virenvshah) [inline minibot/scripts/bot_script.py:2]: Could you make sure bot_script.py doesn't get pushed to GitHub, add this file to the .gitignore

File: minibot/new_minibot.py [hunk 19]
```diff
@@ -94,14 +97,16 @@ def parse_command(cmd, tcpInstance):
...
                 file.close()
-                spawn_script_process(script_name, tcpInstance)
-            except Exception as exception:
-                print("Exception occurred")
-
-    elif key == "STOP":
-        if current_process is not None:
-            current_process.terminate()
-        stop = True
+                """
```
Comment (virenvshah) [inline minibot/new_minibot.py:100]: Use comments like _# Run the Python program_ instead of doc string like comments within functions.

**Review by virenvshah — CHANGES_REQUESTED** [review]

**Comment by virenvshah** [comment]
Looks great now, thanks for making the changes!  I added just a couple more changes and then its good to go from my end!

File: minibot/new_minibot.py (hunk 8 above)
Comment (virenvshah) [reply minibot/new_minibot.py]: Ohhh interesting, we may have to use shared memory to handle it.  No worries, Swathi, Brandon and i can look into when we're cleaning up the backend.  You have done most of the hard work so these small things will be easy to change :).

**Comment by virenvshah** [comment]
Also remember to merge master, there are conflicts in new_minibot.py

File: minibot/scripts/bot_script.py (hunk 18 above)
Comment (ruiqizhu26) [reply minibot/scripts/bot_script.py]: Done!

**Review by virenvshah — APPROVED** [review]

**Review by virenvshah — APPROVED** [review]
LGTM!!!

**Comment by virenvshah** [comment]
Get two more reviewers and we can merge this!

**Review by kabirwalia8300 — APPROVED** [review]

**Review by slawrence100 — APPROVED** [review]
Looks great! Error message concurrency and force-kill button works as expected.

An interesting part I found while testing was that built-in functions will not throw a syntax error on the GUI, but will show up on the server:

Program:
```python
while True:
  bot.moooooooove_forward(100)
```
(no output from the GUI)
Base station terminal output:
```
127.0.0.1 - - [10/Nov/2020 16:47:56] "POST /code/ HTTP/1.1" 500 -
2020-11-10 16:47:56,285 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-11-10 16:47:56,285 INFO sqlalchemy.engine.base.Engine SELECT program.id AS program_id, program.code AS program_code, program.time AS program_time, program.email AS program_email, program.duration AS program_duration 
FROM program
2020-11-10 16:47:56,286 INFO sqlalchemy.engine.base.Engine ()
2020-11-10 16:47:56,286 INFO sqlalchemy.engine.base.Engine ROLLBACK
127.0.0.1 - - [10/Nov/2020 16:47:56] "GET /program/ HTTP/1.1" 200 -
Code len = 36
<class 'bot.pi_bot.PiBot'>
RUNNING SCRIPT
ERROR:tornado.application:Uncaught exception POST /start (127.0.0.1)
HTTPServerRequest(protocol='http', host='localhost:8080', method='POST', uri='/start', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tornado/web.py", line 1697, in _execute
    result = method(*self.path_args, **self.path_kwargs)
  File "base_station_interface.py", line 172, in post
    self.send_program(bot, value)
  File "base_station_interface.py", line 235, in send_program
    func = function_map[match.group(2)]
KeyError: 'moave_forward'
ERROR:tornado.access:500 POST /start (127.0.0.1) 114.22ms
```
I think this is because the _base station's_ Python throws the error instead of the bot, but I think we can fix that as a separate issue.

**Pull request merged**